### ISSUE TITLE
Suggestion: When introspecting, turn lowercase plural table names into uppercase singular model names using `@@map`

### ISSUE THREAD

**Issue by nikolasburk** [prisma/prisma#2502]
In the Prisma schema, model names are [typically](https://github.com/prisma/prisma2/blob/master/docs/data-modeling.md#naming-models) PascalCased. The main reason for this is consistency since Prisma's base scalar types are all spelled in PascalCase notation. Another (rather subjective) reason is that PascalCasing probably feels more familiar to application developers as that's typically also the notation for defining e.g. classes in code.

When introspecting a DB, I'd suggest to transform table names that are written in snake case to PascalCase, or of it's just lowercase without underscores, just uppercase the first letter of the model name using `@@map`.

### Example

Assume this SQL schema:

```sql
CREATE TABLE posts (
    id SERIAL PRIMARY KEY,
    title character varying(256) NOT NULL,
    content text,
    published boolean NOT NULL DEFAULT false,
    "authorId" integer REFERENCES users(id)
);
CREATE UNIQUE INDEX posts_pkey ON posts(id int4_ops);
CREATE TABLE users (
    id SERIAL PRIMARY KEY,
    name character varying(256),
    email character varying(256) NOT NULL UNIQUE
);
CREATE UNIQUE INDEX users_pkey ON users(id int4_ops);
CREATE UNIQUE INDEX users_email_key ON users(email text_ops);
```

With the current introspection, this results in the following Prisma schema:

```prisma
model posts {
  authorId  users?
  content   String?
  id        Int     @id
  published Boolean @default(false)
  title     String
}

model users {
  email   String  @unique
  id      Int     @id
  name    String?
  postses posts[]
}
```

A few things feel a bit off in that Prisma schema:

- The types of all fields are uppercased except for the relation fields 
- `authorId` is a to-one-relation but links to a pluralized field
- `posts` got one more additional pluralization to `postses`
- The type of `postses` is `posts[]` which seems like it could be a list of a list of posts (basically a 2-dimensional array)

Instead, I suggest that we should try to adjust the introspection such that the following Prisma schema is generated:

 ```prisma
model Post {
  authorId  User?
  content   String?
  id        Int     @id
  published Boolean @default(false)
  title     String

  @@map("posts")
}

model User {
  email String  @unique
  id    Int     @id
  name  String?
  posts Post[]

  @@map("users")
}
```

**Comment by cyrus-za** [comment]
I'd like to see this happening. I tried switching to prsima on a fairly large application and introspection generated 3000+ lines of code that all generates lowercase plurals, which results in nexus generating typescript typings that are duplicated, so the code cannot compile. This is a huge blocker and will take way too much time to manually modify each item

**Comment by janpio** [comment]
@cyrus-za While changing this on Prisma side would certainly work around this problem, this also sounds like a limitation of Nexus that should be fixed. Can you please open an issue on their side as well so it can be considered?

**Comment by janpio** [comment]
Unfortunately not. You will have to introspect your database once, then rename all these models to what you want to have and add `@@map` to them individually. Future Introspection runs (with an existing schema file) should keep those annotations in place.

**Comment by clement-estone** [comment]
> Unfortunately not. You will have to introspect your database once, then rename all these models to what you want to have and add `@@map` to them individually. Future Introspection runs (with an existing schema file) should keep those annotations in place.

That's really unfortunate as the schema is huge and changes frequently, so this means I either have to do a lot of changes manually each time or to write a script that does it for me.

Is there a tracking issue for the feature you mentioned?

**Comment by janpio** [comment]
Why do you think you would have "to do a lot of changes manually each time"? As long as you do not get new tables each time, all your modifications will be kept.

This issue here is kinda in that direction, and the others under the `introspection-configuration` label as well: https://github.com/prisma/prisma/issues?q=is%3Aopen+label%3A%22topic%3A+introspection-configuration%22+sort%3Aupdated-desc

**Comment by clement-estone** [comment]
Oh so the whole schema is not overwritten every time I do a `prisma db pull`? If so that's really great, and this indeed solves my problem once the initial setup is done.

**Comment by clement-estone** [comment]
I think so, it wasn't obvious as usually this kind of command generates a whole new file. Having a mix of both generated code and static code isn't very common, so it could be useful to document it ;)

**Comment by sbusso** [comment]
That would be very useful to have this option and open the compatibility with other conventions.

**Comment by steebchen** [comment]
+1 for this, I think it would be actually even better to go further and transform field names with underscores in the DB (which is a good thing to do, IMO) as follows:

```
CREATE TABLE posts (
    id SERIAL PRIMARY KEY,
    content text,
    published_at boolean NOT NULL DEFAULT false,
    author_id integer REFERENCES users(id)
);
```

should be transformed into this:

```prisma
model Post {
 authorId    User?   @map("author_id")
 content     String?
 id          Int     @id
 publishedAt Boolean @default(false) @map("published_at")

 @@map("posts")
}
```

**Comment by shellscape** [comment]
How has this been a feature request for three years without traction? Competitors such as Hasura and Postgraphile have been doing this out of the box for years.

**Comment by janpio** [comment]
Because it is not such a big deal to do it once by hand, vs. the effort of implementation and the complexity it adds to the product (as this would need to be a configuration option, otherwise it would be a breaking change requiring a major verison etc. etc.). If someone really wants this, they could also implement this via a small script and for example share it with the community by posting here.

**Comment by shellscape** [comment]
@janpio that's not a realistic statement for schemas that are hundreds of lines in length with just as many if not more annotations by hand. that's pretty bad DX by any measure. 

A project here https://github.com/iiian/prisma-case-format is accomplishing the ask in this issue in a relatively small amount of effort, and minimal complexity - which nukes the second half of your argument. A simple "we don't want to do this, that's it" would probably be more genuine.

If I were a member of Prisma and I wanted to be a good steward for the userbase, I would make an offer to the author of that util to integrate.

**Comment by janpio** [comment]
Then please do use https://github.com/iiian/prisma-case-format if it solves your problem. That is exactly why we encourage our community to build such projects, and shared them in the relevant issues.

> A simple "we don't want to do this, that's it" would probably be more genuine.

We want this, but it is not a priority right now, so we do not do it yet. Rest assured we do not sit here twiddling our thumbs, but we work on other things that have more impact in our opinion. Offering @iiian anything would not solve any of the problems I described.

**Comment by vasanth-asokan** [comment]
Found this issue via Ask AI on prisma.io. I could not find an authoritative statement in the docs that @@map would preserve the renames upon a future introspection. (I looked here to begin with: https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/custom-model-and-field-names#using-map-and-map-to-rename-fields-and-models-in-the-prisma-client-api. There's a line at the end that describes that future DB introspections will preserve relational field remappings, but doesn't explicitly mention @@map. Ask AI pointed me to this github thread and the clarification is in the middle of the issue. 

Worth mentioning this directly in the docs. To what @clement-estone said, it's not very common to have a generated file be partly generated and partly editable and so it's not natural to assume so.

**Comment by janpio** [comment]
@vasanth-asokan I created a docs issue based on your comment: https://github.com/prisma/docs/issues/5644

**Comment by vasanth-asokan** [comment]
Thank you @janpio! Much appreciated.
